[FE] refactor: Progressbar 모션이 끝난 후 리다이렉트 되도록 수정

eunsoA opened the pull request:
# #️⃣ Issue Number
#324 

## 🕹️ 작업 내용

한 줄 요약 : Progressbar 모션이 끝난 후 ResultPage로 리다이렉트 되도록 수정했어요

- [x] useProgress 훅을 개선 (commit : 21c0701a3ff815b798857689719c3390579c5ba3)
  - 프로그레스바의 기간, 최소 및 최대 값 상수 추가
  - 애니메이션 로직 개선 및 성능 최적화
  - 매개변수 유효성 검사 추가
  - 프로그레스바 즉시 완료 기능 (complete) 개선
- [x] Progressbar 컴포넌트를 개선 (commit : da5cbbf96a01673107be3b485bbc90e7c7e968fb )
  - useProgress 훅에서 `rAF` 과 Progressbar의 컴포넌트에서 css로 `transition`을 사용하고 있어서 `rAF` 만 사용하도록 수정
- [x] ProgressLoading 컴포넌트 렌더링 최적화를 위한 useLocations 훅 개선 (commit : e8e39873149fb4c63ef396c84577e34dae87fc4a)
  - useLocations 훅에서 getRecommendationId와 getRecommendationResult를 사용하는 getRecommendation 함수 추가
  - useLocation에서 isProgressLoading 상태를 추가하여 ProgressLoading 컴포넌트 렌더링 상태 개선
  - 변경된 useLocation 에 맞춰 MeetingForm의 handleSubmit 로직 변경
  - ProgressLoading 컴포넌트에 isComplete prop 추가 및 로딩 완료 시 상태 관리 개선
  - IndexPage에서 로딩 상태(isProgressLoading)에 따라 ProgressLoading 컴포넌트 변경
  - ResultPage에서 data 가 없는 경우 fetch하도록 로직 변경
- [x] BaseLoading 컴포넌트 렌더링 최적화 (commit : df6ad7f7d997971b3bc000ecac062d7141e8afd8)
  - BaseLoading 컴포넌트에서 로고 렌더링을 StaticContent로 분리하여 메모이제이션 적용하여 코드 성능 최적화

## 📋 리뷰 포인트

- [ ] 기존의 방식은 `indexPage>MeetingForm` 에서 `handleSubmit` 실행 시, `indePage`에서 `getRecommendationId`를 사용해 `id`를 받고, 이 `id`를 통해 `ResultPage`에서 `getRecommendationResult`를 사용해서 `location` 데이터를 받아왔어요. 자연스러운 `ProgressLoading` 컴포넌트의 렌더링을 위해 `getRecommendationId`의 동작과 `getRecommendationResult`의 동작을 일괄로 실행하는 `getRecommendation`를 추가했어요. 이 과정에서 불필요한 로직이나 코드가 있는지 확인해주세요.
- [ ] `ProgressLoading` 컴포넌트의 렌더링 과정에서 `BaseLoading` 컴포넌트의 Logo가 불필요하게 재렌더링이 일어나고 있어서 메모제이션을 통해 재렌더링을 방지했어요. 이에 대해 어떻게 생각하시는지 의견이 궁금해요. 
- [ ] 로컬에서 아래의 의도대로 잘 동작하는지 확인해주세요.

## 📸 스크린샷
![화면 기록 2025-09-11 오후 8 45 24](https://github.com/user-attachments/assets/60412aa1-f454-453b-bdef-895cb0cfc14e)
- 스크린샷의 의도는 아래와 같습니다. 
  - 첫 요청 시 ProgressLoading의 게이지가 100까지 찬 후 ResultPage로 렌더링 된다. 
  - 이후 다시 IndexPage에서 재요청을 보내도 의도대로 ProgressLoading의 게이지가 100까지 찬 후 ResultPage로 렌더링 된다. 
  - 렌더링된 ResultPage에서 새로고침하여도 `url`의 `id`를 통해 location의 데이터를 fetch해 올 수 있다.

## 🔮 기타 사항

- 추후에 브라우저의 localStorage에서 id에 대한 data를 캐싱하여, 성능 최적화 관련 팀 요구사항을 만족할 수 있을 것 같습니다!

- 프로그레스바 애니메이션 구현 개선 방안
  - 기존 프로그레스바의 애니메이션은 requestAnimationFrame(rAF)을 통한 JavaScript 기반 애니메이션과 CSS transition 속성이 중복 적용되어 있었어요. rAF는 JavaScript 레벨에서 progress 상태값을 점진적으로 변경하고, CSS transition은 동일한 width 속성 변화에 대해 또다시 보간(interpolation)을 수행하고 있어 불필요한 이중 애니메이션이 발생해요. 
  - 이를 개선하기 위해 CSS transition을 제거하고 rAF만을 사용하여 애니메이션을 구현하면, 단일 애니메이션 레이어에서 더 정확하고 효율적인 진행률 표시가 가능해요. 
  - 이는 브라우저의 렌더링 최적화에 도움을 주며, 특히 저사양 디바이스에서 더 부드러운 애니메이션 효과를 기대할 수 있어요.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **새로운 기능**
  * 글로벌 진행 상태(isProgressLoading) 및 진행 완료 플래그(isReadyToComplete) 도입.
  * 추천 호출이 전체 응답({id, data})을 반환하도록 확장되어 호출 결과 사용 방식이 변경됨.

* **리팩터**
  * 진행 애니메이션을 RAF 기반으로 재작성하고 완료 트리거(complete) 추가.
  * 페이지 로딩/오류 처리 흐름 정리 및 로딩 컴포넌트의 정적 콘텐츠를 메모이제이션 처리.

* **스타일**
  * 진행 막대의 너비 애니메이션 제거.

* **테스트**
  * 추천 결과 테스트에서 반환 데이터 검증 방식 조정.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

eunsoA committed:
feat: useProgress 개선

- 프로그레스바의 최소 및 최대 값 상수 추가
- 애니메이션 로직 개선 및 성능 최적화
- 매개변수 유효성 검사 추가
- 프로그레스바 즉시 완료 기능 (complete) 개선

eunsoA committed:
fix: 프로그레스바 애니메이션 제거

- 프로그레스바의 width 변화에 대한 애니메이션 효과를 제거하여 성능 최적화

eunsoA committed:
refactor: ProgressLoading 컴포넌트 렌더링 최적화를 위한 useLocations 훅 개선

- useLocations 훅에서 getRecommendationId와 getRecommendationResult를 사용하는 getRecommendation 함수추가
- useLocation에서 isProgressLoading 상태를 추가하여 ProgressLoading 컴포넌트 렌더링 상태 개선
- 변경된 useLocation 에 맞춰 MeetingForm의 handleSubmit 로직 변경
- ProgressLoading 컴포넌트에 isComplete prop 추가 및 로딩 완료 시 상태 관리 개선
- IndexPage에서 로딩 상태(isProgressLoading)에 따라 ProgressLoading 컴포넌트 변경
- ResultPage에서 data 가 없는 경우 fetch하도록 로직 변경

eunsoA committed:
refactor: BaseLoading 컴포넌트 렌더링 최적화

- BaseLoading 컴포넌트에서 로고 렌더링을 StaticContent로 분리하여 메모이제이션 적용
- 코드 성능 최적화

eunsoA committed:
refactor: useLocations 훅 lint 적용하여 코드 가독성 향상

eunsoA committed:
refactor: animationRef의 타입을 null을 허용하도록 수정

eunsoA commented on frontend/src/features/progressbar/hooks/useProgress.ts:
```diff
@@ -1,60 +1,113 @@
+  targetProgress = PROGRESS_MAX_VALUE
+}: UseProgressParams = {}): UseProgressReturns => {
   const [progress, setProgress] = useState(initialProgress);
+  const animationRef = useRef<number>(null);
```
요거 반영했어요 (commit : [3cd5ef2](https://github.com/woowacourse-teams/2025-moitz/pull/327/commits/3cd5ef2a009735a8255141b5cc7b6037adc1e4cb) )

eunsoA committed:
fix: IndexPage의 data 접근 시 null 체크 추가

- recommendedLocations 접근 시 옵셔널 체이닝 적용하여 안전성 개선

eunsoA commented on frontend/src/pages/components/indexPage/IndexPage.tsx:
```diff
@@ -11,11 +11,21 @@ import { flex, grid_padding, scroll } from '@shared/styles/default.styled';
+    );
+
+  if (isProgressLoading) {
+    return <ProgressLoading isComplete={isProgressLoading && data.recommendedLocations.length > 0} />;
```
이것도 반영했습니닷 (commit : f5a78e9)

kaori-killer commented:
@coderabbitai  
You are a senior React engineer.

# Task
From this pull request’s code, identify **exactly one place** where the code is written in an imperative style, and explain how to make it **more declarative**.

# How to respond
Start your response with the following markdown title:  
`# 1. ✨ PR 내용 요약`

Then respond with:
- **Imperative pattern name or symptom**
- **Where it appears** (files + line ranges or clear examples)
- **Why it’s not declarative**
- **AS-IS**: _Show the current (imperative) code snippet_
- **TO-BE**: _Show the rewritten (declarative) version_
- **Why this change improves React code quality** (≤ 2 sentences)

# Style
Respond in bullet points, and include only one clear improvement point.

# Language
Please write your response **in Korean**.

kaori-killer commented:
@coderabbitai  
You are a senior React engineer.

# Task
해당 PR 내용을 바탕으로 **PR 본문을 요약해서** 주세요.

# How to respond
Start your response with the following markdown title:  
`# 1. ✨ PR 내용 요약`

Then respond with:
- **한 줄 요약**
- **주요 변경 사항 목록**
- **리뷰어가 확인해야 할 포인트 (2~3개)**

# Style
- 간결하고 읽기 쉬운 불릿 포인트로 정리
- 필요하면 코드/스크린샷 대신 설명만 사용
- 전체 분량은 10줄 이내

# Language
Please write your response **in Korean**.

kaori-killer commented on frontend/src/entities/location/hooks/useLocations.ts:
```diff
@@ -64,13 +69,37 @@ const useLocations = (): useLocationsReturn => {
+        setIsProgressLoading(false);
+      }
+    },
+    [],
```
@eunsoA 

의존성 배열을 채워주실 수 있나요~?

[리액트 공식문서 - 반응형값 의존성 지정](https://ko.react.dev/reference/react/useEffect#specifying-reactive-dependencies)에서도 의존성을 비워두는 것보다 지정하는 것을 권장해서요 🙌

eunsoA replied:
확인하고 수정했습니다 :) (commit : 5fc68e1e2f0f452e152ff0bf37b9a8b4ec601d02)

kaori-killer commented on frontend/src/entities/location/hooks/useLocations.ts:
```diff
@@ -7,13 +7,17 @@ import { Location } from '@entities/location/types/Location';
 
 export type useLocationsReturn = {
   data: Location;
+  isProgressLoading: boolean;
   isLoading: boolean;
   isError: boolean;
   errorMessage: string;
   getRecommendationId: (
     requestBody: RecommendationRequestBody,
   ) => Promise<string>;
-  getRecommendationResult: (id: string) => Promise<void>;
+  getRecommendationResult: (id: string) => Promise<Location>;
+  getRecommendation: (
+    requestBody: RecommendationRequestBody,
+  ) => Promise<{ id: string; data: Location }>;
 };
```
`useLocations` 내부에 상태가 점점 늘어나면서 관리가 어려워지고 있네요 😭

eunsoA replied:
맞아요. 현재 useLocations 내부에서 여러 상태와 복잡한 비동기 플로우를 관리하고 있어서 확실히 어지러워 보여요

대안으로는 데이터 페칭 로직을 별도 서비스 레이어로 분리하는 방법이 있을 것 같은데요. 이 부분은 현재 테스크 범위를 벗어나는 것 같아 우선 헤일리의 의견을 묻고 싶었어요. 어떻게 생각하시나요?

eunsoA replied:
코어 타임에 이야기했던 대로, 우선은 이슈에 추가해두었어요! [이슈 바로가기](https://github.com/woowacourse-teams/2025-moitz/issues/372) :>

kaori-killer commented on frontend/src/features/loading/components/baseLoading/BaseLoading.tsx:
```diff
@@ -9,6 +9,14 @@ interface BaseLoadingProps {
+  </div>
+));
+
+StaticContent.displayName = 'StaticContent';
```
어떤 걸 확인하기 위해서 displayName을 지정했나요?!

eunsoA replied:
이건 성능 최적화를 위한 조치였어요. ProgressLoading에서 프로그레스바가 애니메이션될 때마다 progress 상태가 빈번하게 변경되면서 BaseLoading 전체가 리렌더링되거든요.

그런데 로고는 정적 콘텐츠라서 매번 다시 렌더링될 필요가 없어서 memo()로 최적화했습니다. 그리고 memo()로 감싼 익명 함수 컴포넌트는 ESLint 규칙상 `displayName`이 필요해서 추가했어요 !

실제로 프로그레스바 애니메이션 중에는 초당 60번 정도 리렌더링이 일어나니까, 로고 최적화가 필요하다고 생각했어요.

kaori-killer commented on frontend/src/features/loading/components/progressLoading/ProgressLoading.tsx:
```diff
@@ -1,12 +1,24 @@
+function ProgressLoading({ isComplete }: ProgressLoadingProps) {
+  const { progress, complete } = useProgress({ initialProgress: 0, targetProgress: 90 });
+
+  useEffect(() => {
+    if (isComplete && progress < 100) {
+      complete();
+    }
+  }, [isComplete]);
```
@eunsoA 여기도 의존성 배열을 변경하면 어떨까요?!

eunsoA replied:
해당 부분의 의존성 배열에 대해 저도 동일하게 고민했으나 추가하지 않았습니다. 이유는 아래와 같습니다!

---
### 변경하지 않은 이유
ProgressLoading 컴포넌트는 2개의 fetch를 순차적으로 요청하는 동안 프로그래스바를 렌더링하며, 응답을 정상적으로 받으면 애니메이션을 종료한 뒤 언마운트됩니다.

`isComplete`는 fetch가 완료되어 프로그래스바 렌더링을 종료해도 됨을 의미하므로, **isComplete가 false에서 true로 변경되는 시점에만** complete 함수를 실행하도록 구현했습니다.

이때, 의존성 배열에 progress를 추가할 경우, 아래와 같은 문제가 발생합니다.
- progress 값이 0부터 100까지 증가하는 동안 useEffect가 계속 실행됨
- 매번 `isComplete && progress < 100` 조건을 검사하는 불필요한 연산 발생
- 실제로는 isComplete가 true가 되는 시점에만 실행되면 충분함

위와 같은 이유로 의존성 배열에 `isComplete`만 추가하였습니다.

---

### 고민해보았던 대안

React의 규칙을 따르기 위해, 아래와 같은 방법도 있을 것 같습니다.

```js
// 대안 : useCallback 사용
const handleComplete = useCallback(() => {
  if (progress < 100) {
    complete();
  }
}, [progress, complete]);

useEffect(() => {
  if (isComplete) {
    handleComplete();
  }
}, [isComplete, handleComplete]);
```
하지만 현재 로직상 isComplete 변경 시점에만 실행하면 되므로, 기존 구현이 가장 적합하다고 판단했습니다.

---

### 질문
이 외의 더 나은 해결책이 있을지에 대한 헤일리의 의견이 궁금합니다 !! 👀👀

eunsoA replied:
대면 피드백에서 논의한대로, progress<100 을 판단하는 로직을 지우고 의존성 배열에 complete 을 추가했어요 ~! (commit : 28ffd406b4c2ea4d551c6a6ef3b22af989a626f6)

kaori-killer commented on frontend/src/features/loading/components/progressLoading/ProgressLoading.tsx:
```diff
@@ -1,12 +1,24 @@
+function ProgressLoading({ isComplete }: ProgressLoadingProps) {
+  const { progress, complete } = useProgress({ initialProgress: 0, targetProgress: 90 });
+
+  useEffect(() => {
+    if (isComplete && progress < 100) {
+      complete();
+    }
+  }, [isComplete]);
```
```tsx
(isComplete && progress < 100)
```

위 구문이 어떤 의도인지 헷갈리네요 🤔

조건에서 `isComplete` 값이 참이면서
`progress` 값이 100 미만이어야 하는 이유가 무엇일까요?
`isComplete` 값만 참이면 안되는 걸까요?

eunsoA replied:
`isComplete`와 `progress < 100`은 서로 다른 상태를 나타냅니다!
네이밍이 비슷한 맥락으로 보여서 헷갈릴 수 있지만, 각각 다른 책임을 가지고 있어요.

---
### 두 상태의 역할 분리

progress (UI 상태)
- useProgress 훅에서 관리하는 **애니메이션 진행률**
- 0 → 90까지 자동 증가 후 대기 (targetProgress: 90)
- complete() 호출 시 90 → 100으로 완료 애니메이션 실행

isComplete (비즈니스 상태)
- **API 요청 완료 여부**를 나타내는 외부 상태
- 두 개의 fetch가 모두 완료되면 true로 변경
- 프로그레스바 완료 신호 역할
---

### 조건문이 필요한 이유

```ts
if (isComplete && progress < 100) {
  complete(); // 90 → 100 완료 애니메이션 실행
}
```

시나리오별 동작:
1. **API가 빨리 완료된 경우**: progress가 90 미만일 때 isComplete가 true → complete() 실행
2. **API가 늦게 완료된 경우**: progress가 90에서 대기 중일 때 isComplete가 true → complete() 실행  
3. **이미 완료된 경우**: progress가 100이면 complete() 실행하지 않음 (중복 방지)

`progress < 100` 조건이 없다면 이미 완료된 상태에서도 불필요하게 complete()가 호출될 수 있어요.

---
### 리뷰 반영하여 네이밍 개선 제안
혼동을 줄이기 위해 다음과 같은 네이밍으로 변경했어요. 어떤가요 ? 👀
- `isComplete` → `isReadyToComplete ` 변경 (commit : 18f1a1b3fdc0f6fbb040c9a70f514c22184d38cb)

kaori-killer commented on frontend/src/entities/location/hooks/useLocations.ts:
```diff
@@ -64,13 +69,37 @@ const useLocations = (): useLocationsReturn => {
     isLoading,
     isError,
     errorMessage,
     getRecommendationId,
     getRecommendationResult,
+    getRecommendation,
```
세 개를 하나로 통합할 수는 없을까요? 

특히 `getRecommendation`는 `getRecommendationId`과 `getRecommendationResult` 네이밍이 비슷해서 어떤 데이터인지 헷갈릴 것 같아요!

eunsoA replied:
좋은 의견 감사해요! 네이밍이 헷갈릴 수 있다는 점에 동의해요 👍

`getRecommendationId` 와 `getRecommendationResult` 두 개의 함수를 `getRecommendation`으로 통일하는 것도 고민했지만, 결국 필요해서 남겨두었습니다.

---
### 이유
현재 구조는 아래와 같아요:
- getRecommendation: 전체 플로우 (ID 요청 → 결과 조회)
- getRecommendationResult: 결과만 조회 (URL 직접 접근용)
- getRecommendationId: 단독으로는 사용하지 않음

아시다시피 사용자 시나리오는 두가지가 있어요:
1. IndexPage → ResultPage: getRecommendation (프로그레스 로딩)
2. URL 직접 접근: getRecommendationResult (일반 로딩)

따라서 url로 직접 접근 로직에서 필요했기에, 세가지를 하나로 통합할 수 없었어요.

추가로, `getRecommendationId`는 단위 테스트를 위해서 남겨두었어요.
위의 코멘트에서 남겼던 대로, 개선이 필요해보여서 다음에 함께 논의해보면 좋겠습니다 : >

---

### 개선점
말씀해주신대로 네이밍을 아래와 같이 수정했어요.
- `getRecommendation` -> `getRecommendationFull` (commit : 035f2e3f48033db61b7f0d955c6e2fe066a138ae)

이렇게 하면 용도가 더 명확해질 것 같아요. 어떻게 생각하시나요?

kaori-killer commented on frontend/src/features/progressbar/hooks/useProgress.ts:
```diff
@@ -1,57 +1,121 @@
-    const animationFrame = () => {
-      const currentTime = Date.now();
-      const elapsed = currentTime - startTime;
+  /**
+   * 프로그레스바의 애니메이션을 처리하는 함수
+   * @param duration - 애니메이션 지속 시간 (밀리초)
+   * @param from - 시작 진행률
+   * @param to - 목표 진행률
+   */
```
주석 남겨주셔서 감사합니다 👍

이렇게 계산 로직 위주의 함수는 주석이 코드를 읽을 때 도움되더라고요!

kaori-killer commented on frontend/src/features/progressbar/hooks/useProgress.ts:
```diff
@@ -1,57 +1,121 @@
+        const elapsed = currentTime - startTimeRef.current;
+
+        if (elapsed >= duration) {
+          return to;
+        }
 
-      setProgress(easedProgress);
-      animationId = requestAnimationFrame(animationFrame);
+        const ratio = elapsed / duration;
+        const easedProgress =
+          startProgress + (to - startProgress) * (1 - Math.pow(1 - ratio, 3));
+        const easedProgressNumber = Math.floor(easedProgress * 1000) / 1000;
+        return easedProgressNumber;
+      });
+
+      animationRef.current = requestAnimationFrame(animationFrame);
     };
 
-    animationId = requestAnimationFrame(animationFrame);
+    animationRef.current = requestAnimationFrame(animationFrame);
+  }, []);
```
계산 로직이 바뀐 것 같은데 해당 부분을 설명해주시거나 PR에 적어주실 수 있나요?! 🙌

eunsoA replied:
PR에 설명을 포함하면 본문이 더 길어질 것 같아서 .. 😅 코멘트에 남겨 놓을게요! 그리고 필요하다면 요청해주신대로 대면으로 다시 설명드릴게요 : >

---
### 변경 이유
기존 로직은 0에서 시작하는 애니메이션만 지원했는데, 이제 임의의 진행률에서 시작할 수 있도록 개선했습니다.

### 이전 (0에서만 시작 가능)
```ts
const ratio = elapsed / duration;
const easedProgress = Math.min(
  targetProgress, targetProgress * (1 - Math.pow(1 - ratio, 3))
);
```
- 항상 0에서 targetProgress로만 애니메이션
- complete() 함수에서 현재 진행률에서 100%로 이어서 애니메이션하기 어려움

### 이후 (임의 지점에서 시작 가능)
```ts
const ratio = elapsed / duration;
const easedProgress =
  startProgress + (to - startProgress) * (1 - Math.pow(1 - ratio, 3));
const easedProgressNumber = Math.floor(easedProgress * 1000) / 1000;
```
- startProgress에서 to까지 자유롭게 애니메이션
- 예: 30%에서 100%로, 0%에서 90%로 등 유연한 애니메이션 지원
- `Math.floor(easedProgress * 1000) / 1000`는 소수점 3자리까지 정밀도 보장으로 부드러운 애니메이션

---
이 변경으로 complete() 함수가 현재 진행률에서 자연스럽게 100%까지 이어서 애니메이션할 수 있도록 했어요.

kaori-killer reviewed (changes requested):
안녕하세요, 클레어 😄 레벨 4 첫 리뷰네요~ 같이 열심히 해봅시다 ㅎㅎ

프로그래스 바 이제 안 멈추고 잘 올라가더라고요 💯 넘 예뻐요!

### ✅ 기능 동작 확인

- [x] IndexPage → ResultPage 이동 시 프로그래스 바가 정상적으로 동작합니다.  
- [ ] ResultPage를 직접 접근했을 때 프로그래스 바가 정상적으로 동작합니다. 

![1](https://github.com/user-attachments/assets/2e345710-b93a-4988-aa92-e9407a37b21f)

특히 **Slow 3G** 상태에서 테스트해보니, PR 본문에 적힌 

> 렌더링된 ResultPage에서 새로고침하여도 url의 id를 통해 location의 데이터를 fetch해 올 수 있다.

이 부분이 실제로는 동작하지 않는 것 같아요. 한 번 더 확인 부탁드려요 🙏

---

### ✨ PR 본문 피드백

PR을 꼼꼼하게 작성해주셔서 리뷰하기 훨씬 수월했어요. 감사합니다.
다만, 읽으면서 느낀 점 몇 가지 개선 점이 있는데요, 공유드릴게요. 편하게 반영해주세요~

1.  본문에는 핵심 흐름만 간단히 적고, 세부적인 기술 설명은 기타 섹션이나 별도 문서로 분리해주시면 가독성이 훨씬 좋아질 것 같아요. 지금은 `작업 내용/리뷰 포인트/기타 사항`에 모두 많은 내용이 들어가 있어서 어디에 집중해야 할지 조금 헷갈렸고 😭 전체를 읽는 데도 시간이 꽤 걸렸습니다 ㅠㅜ (줄여줘~)

2. 사용자 관점에서 먼저 설명해주시면 어떨까요? `rAF`, `transition` 같은 기술 용어보다는 "사용자 입장에서 어떤 개선이 이루어지는지"를 먼저 적어주고, 기술적인 배경은 별도 섹션에 정리해주면 이해가 더 쉬울 것 같아요. 또한, 왜 해당 기술을 선택했는지 간단히 덧붙여주시면 리뷰할 때 도움이 될 것 같습니다! (프로그래스 바 갑자기 왜 잘 동작하는지도 ㅋㅋ)

3. 리뷰 포인트에 로직 관련 확인 요청을 적어주신 것도 좋지만, 직접 코드 레벨에서 코멘트를 남겨주시면 더 도움이 될 것 같아요~! `getRecommendationResult`, `ProgressLoading` 등 글로 보니까 계속 코드 레벨에서 찾으면서 이해해야 하더라고요~

---

이번 PR을 완전히 이해하지 못해서 리뷰가 조금 추상적일 수도 있어요 😅  
혹시 코드 리뷰를 라이브로 진행해서 설명 한번 해주실 수 있을까요? 🥹❤️

eunsoA committed:
refactor: useProgress 훅의 파라미터 유효성 검사로직 분리

eunsoA committed:
refactor: useLocations 훅의 의존성 배열에 getRecommendationId와 getRecommendationResult 추가

eunsoA committed:
refactor: ProgressLoading 컴포넌트의 isComplete 변수 네이밍을isReadyToComplete으로 변경

eunsoA committed:
refactor: getRecommendation 함수의 이름을 getRecommendationFull로 변경

eunsoA commented:
안녕하세요 헤일리자몽 🔥

위의 코멘트 모두 확인 후 답변 달아 놓았습니다.
꼼꼼하게 리뷰해주셔서 감사해요! 항상 덕분에 많이 배우고 있습니닷 🙇‍♂️

---
### 🌀 기능 동작
> 기능 동작 확인
[ ] ResultPage를 직접 접근했을 때 프로그래스 바가 정상적으로 동작합니다.
특히 Slow 3G 상태에서 테스트해보니, PR 본문에 적힌
렌더링된 ResultPage에서 새로고침하여도 url의 id를 통해 location의 데이터를 fetch해 올 수 있다.
이 부분이 실제로는 동작하지 않는 것 같아요. 한 번 더 확인 부탁드려요 🙏

동작하지 않는 게 아니라 로딩에 시간이 걸리는 것 같아요. 우선 제 로컬에서 3G로 테스트시 느리지만 정상적으로 데이터를 받아오고 있기에 다시 한번 확인해주실 수 있나요 ? 👀

https://github.com/user-attachments/assets/fc1d651f-3ed1-4838-bd2a-9c0e01e75c15

---
### 🌀 PR 본문 피드백 반영

PR 본문에 대해 피드백 주셔서 감사해요! 연관된 변경사안을 모두 설명하려다보니 PR의 양이 많아졌던 것 같네요 .. 🥲 가독성을 놓친 점, 죄송합니다. 비교를 위해 상단의 PR을 수정하지 않고, 아래에 다시 수정해서 PR을 작성해봤어요.
추가로 개선할 부분이 있다면 말씀해주세요 : >

또한 리뷰 포인트도 확인부탁드려요!

eunsoA commented:
# #️⃣ Issue Number
#324 

## 🕹️ 작업 내용

배경 : Progressbar의 애니메이션이 완료되기 전에 ResultPage로 리다이렉트되거나, bar 컴포넌트가 정상적으로 렌더링 되지 않는 문제가 있었어요.
한 줄 요약 : Progressbar 애니메이션이 100% 완료된 후 ResultPage로 리다이렉트되도록 UX를 개선했어요.

- [x] 프로그래스바 애니메이션이 자연스럽게 진행되도록 수정했어요 (commit : 21c0701a3ff815b798857689719c3390579c5ba3, da5cbbf96a01673107be3b485bbc90e7c7e968fb )
- [x] ProgressLoading 컴포넌트 렌더링 최적화를 위한 useLocations 훅을 개선했어요 (commit : e8e39873149fb4c63ef396c84577e34dae87fc4a)
- [x] BaseLoading 컴포넌트에서 로고 렌더링을 최적화했어요 (commit : df6ad7f7d997971b3bc000ecac062d7141e8afd8)

## 📋 리뷰 포인트

- [ ] `isProgressLoading` 상태를 위해 `useLocation` 훅에 `getRecommendationId`의 동작과 `getRecommendationResult`의 동작을 통합하여 순차적으로 실행하는 `getRecommendation`를 추가했어요. 이 과정에서 불필요한 로직이나 코드가 있는지 확인해주세요.
- [ ] `ProgressLoading` 컴포넌트의 렌더링 과정에서 `BaseLoading`의 Logo 컴포넌트가 불필요하게 재렌더링이 일어나고 있어서 메모제이션을 통해 재렌더링을 방지했어요. 이에 대해 어떻게 생각하시는지 의견이 궁금해요.
- [ ] 로컬에서 아래의 사용자 시나리오대로 잘 동작하는지 확인해주세요.
  - 첫 fetch 요청 시 ProgressLoading의 애니메이션이 100% 실행된 후 ResultPage로 렌더링 된다. 
  - 렌더링된 ResultPage에서 새로고침하여도 `url`의 `id`를 통해 location의 데이터를 fetch해 올 수 있다. 이때, 로딩화면은 `BaseLoading`이 보여진다.

## 📸 스크린샷
두가지 사용자 시나리오:
1. IndexPage → ResultPage: getRecommendation (프로그레스 로딩)
2. URL 직접 접근: getRecommendationResult (일반 로딩)

| Before (1번 경로 접속) | Before (2번 경로 접속) | After (1번 경로 접속) | After (2번 경로 접속) |
|--------|--------|--------|--------|
| ![화면 기록 2025-09-24 오전 4 54 52](https://github.com/user-attachments/assets/9aa2aa9f-b3ba-44b4-908d-9d9288102b16) | ![화면 기록 2025-09-24 오전 4 55 17](https://github.com/user-attachments/assets/ac90ff55-0b61-462d-be52-d76dcee93e09) | ![화면 기록 2025-09-24 오전 4 55 49](https://github.com/user-attachments/assets/5fc5d86f-33a9-4ada-a25b-b7635ce801b3) | ![화면 기록 2025-09-24 오전 4 56 06](https://github.com/user-attachments/assets/2031b16a-a978-4bdd-822c-e473ab93fd97) |

eunsoA committed:
refactor: ProgressLoading 컴포넌트의 로딩 완료 조건 수정

- isReadyToComplete가 true일 때만 complete() 호출하도록 변경하여 로직 간소화
- 의존성 배열에 complete 추가

eunsoA committed:
refactor: IndexPage의 FallBackPage의 reset 함수에서 window.location.reload() 호출을 제거
